File: src/main/java/self/ed/main/RunReducer.java
package self.ed.main;

import static java.lang.System.currentTimeMillis;

import static self.ed.main.DataDirs.READY_DIR;
import static self.ed.main.DataDirs.REDUCER_FAILED_DIR;
import static self.ed.main.DataDirs.REDUCER_FIXED_DIR;
import static self.ed.util.FileUtils.appendFile;
import static self.ed.util.FileUtils.listFiles;
import static self.ed.util.FileUtils.readFile;
import static self.ed.util.FileUtils.writeFile;
import static self.ed.util.SudokuUtils.countOpen;
import static self.ed.util.SudokuUtils.fromString2D;
import static self.ed.util.SudokuUtils.toString1D;
import static self.ed.util.SudokuUtils.toString2D;

import java.io.File;
import java.nio.file.Path;
import java.util.List;
import java.util.concurrent.atomic.AtomicLong;

import self.ed.generator.Reducer;

public class RunReducer {

  public static void main(String[] args) {
    reduceFailedByGeneratorTimeout();
//    copyReducedToReady();
  }

  private static void reduceFailedByGeneratorTimeout() {
    Reducer reducer = new Reducer();
    AtomicLong minimizedCount = new AtomicLong();
    List<File> files = listFiles(REDUCER_FAILED_DIR);
    for (File file : files) {
      System.out.println(file.getName());
      Integer[][] input = fromString2D(readFile(file));
      long startTime = currentTimeMillis();
      Integer[][] output = reducer.reduce(input);
      System.out.println("Time: " + (currentTimeMillis() - startTime) / 1000d + "s");
      long inputCount = countOpen(input);
      long outputCount = countOpen(output);
      String outFile = outputCount + "-" + file.getName().split("-", 2)[1];
      writeFile(REDUCER_FIXED_DIR.resolve(outFile), toString2D(output));
      if (outputCount != inputCount) {
        minimizedCount.incrementAndGet();
        System.out.println("Minimized " + file.getName() + ":" + inputCount + " => " + outputCount);
      }
      file.delete();
    }
    System.out.println("Minimized " + minimizedCount.get() + " of " + files.size());
  }

  private static void copyReducedToReady() {
    List<File> files = listFiles(REDUCER_FIXED_DIR);
    for (File file : files) {
      System.out.println(file.getName());
      Integer[][] result = fromString2D(readFile(file));
      Long openCount = countOpen(result);
      Path readyFile = READY_DIR.resolve(openCount + ".txt");
      appendFile(readyFile.toFile(), toString1D(result) + "\n");
      file.delete();
    }
  }
}

File: cpp-sudoku/src/run_reducer.cpp
#include "run_reducer.hpp"
#include "reducer.hpp"
#include "utils.hpp"
#include "data_dirs.hpp"
#include <iostream>
#include <chrono>
#include <atomic>

namespace sudoku {

void RunReducer::reduceFailedByGeneratorTimeout() {
    Reducer reducer;
    std::atomic<long> minimizedCount{0};
    std::vector<std::filesystem::path> files = Utils::listFiles(DataDirs::REDUCER_FAILED_DIR);
    
    for (const auto& file : files) {
        std::cout << file.filename().string() << std::endl;
        std::string content = Utils::readFile(file);
        auto input = Utils::fromString2D(content);
        
        auto startTime = std::chrono::steady_clock::now();
        auto output = reducer.reduce(input);
        auto endTime = std::chrono::steady_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(endTime - startTime);
        
        std::cout << "Time: " << duration.count() / 1000.0 << "s" << std::endl;
        
        long inputCount = Utils::countOpen(input);
        long outputCount = Utils::countOpen(output);
        
        std::string filename = file.filename().string();
        size_t dashPos = filename.find('-');
        std::string outFile = std::to_string(outputCount) + "-" + filename.substr(dashPos + 1);
        
        Utils::writeFile(DataDirs::REDUCER_FIXED_DIR / outFile, Utils::toString2D(output));
        
        if (outputCount != inputCount) {
            minimizedCount++;
            std::cout << "Minimized " << filename << ":" << inputCount << " => " << outputCount << std::endl;
        }
        
        std::filesystem::remove(file);
    }
    
    std::cout << "Minimized " << minimizedCount.load() << " of " << files.size() << std::endl;
}

void RunReducer::copyReducedToReady() {
    std::vector<std::filesystem::path> files = Utils::listFiles(DataDirs::REDUCER_FIXED_DIR);
    
    for (const auto& file : files) {
        std::cout << file.filename().string() << std::endl;
        std::string content = Utils::readFile(file);
        auto result = Utils::fromString2D(content);
        
        long openCount = Utils::countOpen(result);
        std::filesystem::path readyFile = DataDirs::READY_DIR / (std::to_string(openCount) + ".txt");
        Utils::appendFile(readyFile, Utils::toString1D(result) + "\n");
        
        std::filesystem::remove(file);
    }
}

} // namespace sudoku
